Paper: A Real-Time 3D Audio Simulator for Cognitive Hearing Science

mwickert opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

Commit:
Practice Using last year's paper

Commit:
Merge branch '2019' of https://github.com/scipy-conference/scipy_proceedings into 2019

Commit:
Bring in correct figures

Commit:
Updated paper content.

Commit:
Updated figures and text.

deniederhut commented:
Hey @mwickert ! There are a lot of sections in your paper that have something like this:

> More writing TBD. 

Did you forget to push some commits with the new content?

mwickert commented:
I am working on it now and hope to have more gaps filled later today, and hopefully all completed Sunday evening.

Is this reasonable by you?

deniederhut commented:
Great! Just as a reminder, papers are supposed to have been complete by the submission deadline, which was a couple of weeks ago.
On Jun 8, 2019, 09:47 -0400, Mark Wickert <notifications@github.com>, wrote:
> I am working on it now and hope to have more gaps filled later today, and hopefully all completed Sunday evening.
> Is this reasonable by you?
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

Commit:
Modify figures and fill in missing text. Build errors hampering progress.

Commit:
Build error resolved for now. Continue to write.

Commit:
More writing.

Commit:
Getting closer to writing complete.

Commit:
Final text additions to get paper ready for review.

Commit:
Minor text and figure changes. Code details for the last section in prep.

senolpiskin reviewed (changes requested):
Hey Mark, 

I am reviewing this paper for SciPy 2019. After a quick review, I think:

PDF files within the RST file are not viewed automatically.
Some of the links are not functioning correctly.
Youtube video is not available.

Could you please check these issues so that I can review extensively.

Regards,
Senol

mwickert commented:
Senol,

Working from the bottom up, I would not expect Youtube video to do anything at this point.

When you say links not functioning do you mean in the paper itself?

Finally as for viewing pdf file, I am not sure what else I need to do. I wrote two papers for the 2018 Scipy and I am using the same computer and GitHub configuration as I did in the past. The paper builds to pdf fine on my local machine. I assume that when I push to GitHub the build server should take over and build the pdf of the paper on the server.

Maybe Dillon can comment.

Mark

deniederhut commented:
Correct - the youtube link is a placeholder for when we have the recorded talk online after the conference.

As for rendering, @senolpiskin you can see the paper here http://procbuild.scipy.org/download/mwickert-2019 on the buildserver

senolpiskin commented:
Yes, I can see now. Thanks. I will send my comments soon.

senolpiskin commented:
Here is my review report:

The manuscript describes a 3D audio simulator that can be used in real-time applications for cognitive hearing science. The proposed software can be useful in many aspects of hearing applications. It is within the scope of the SciPy conference and the audience can be highly interested in reading the paper and using the corresponding codes. But, the paper needs several major changes as I have noted below:

The first concern is the language of the manuscript: There are many spelling and grammar mistakes that should be addressed. Furthermore, some statements are not clear. I have mentioned only a few of them although there are many others:

Please rephrase: "To that end this simulator
is integral to planned human subject testing where headphone
testing will be used to blind test subjects for their perception of
where a sound source is located."

Use a comma and rephrase: "To produce a synthesized 3D audio sound field starts with a
geometry."

Not clear and there are two "the": " For a given source location
(x, y,z) = (x1,−x2, x3) pointing at the the origin, we transform
from rectangular coordinates to cylindrical coordinates as shown
in Figure 1."

Please use more formal English: "This transformation is motivated by [Fitzpatrick],
and will be explained more fully in a later section. "
For instance, you could say "This transformation is motivated by [Fitzpatrick],
and explained in detail in section Section Name'."

Do you mean 'mentioned'?: "Note also the
second rectangular coordinate frame mention above is the notation
used by CIPIC."

Not clear, please check again and rephrase: "The 3D audio rendering provided by the simulator developed
in this paper, relies on the 1250 HRIR measurements were taken
using the geometrical configuration shown in Figure 2."

Please remove 'The': "For a source on the 1 m reference sphere, or further away, the there is no ..."

Please rephrase: "When the source is inside is the unit sphere sound parallax ..."

Please check for more. There are many others. I suggest a native English speaker revise the whole text.


Other than the language issues, I have some other suggestions. Please see below:

The abstract is repeated in the introduction section of the main text. That is unnecessary. Please remove that part from the introduction.

Define CIBIC abbreviation before its first usage. 

Please define the symbols of the mathematical variables such as azimuth angle, elevation,... in the text at the first usage.

What does it mean? "For a given source location
(x, y,z) = (x1,−x2, x3) pointing at the the origin" 
Please explain the equation in more detail. It is not very clear.

Use the same terminology throughout the whole manuscript: "we transform
from rectangular coordinates to cylindrical coordinates" and "convert from cylindrical coordinates
to cartesian"
Use either cartesian or rectangular, not both.

It is not clear what kind of analysis is performed in the section 'Spherical Head Model as a Simple Reference HRIR'. Are you reproducing any previous data here? Or comparing the results with some analytical solutions? If so, what is the error? If not, I suggest performing an error analysis of the results compared to some analytical solution.

I am suggesting re-review after these changes.

mwickert commented:
Senolpiskin,

Thank you for detailed comments. I appreciate the time you have taken to go through my writing, which I acknowledge was not proofed very well. I fact there are details in the spherical head model section that I feel must be added. 

I welcome your re-review. I hope to have a response late tomorrow or early Thursday.

Commit:
Making progress with the spherical head model section.

Commit:
Progress, but details remain in the CIPIC-like database calculation before sweeping from the top of the paper to incorporate the comments of Senolpiskin.

Commit:
Many updates; sphereical head section rework complete. 3D coordinate section updated.

Commit:
The conclusions section is updated with more words.

Commit:
Near final release for the second round of review.

Commit:
Added some more changes to answer reviewer critque.

mwickert commented:
@senolpiskin and @deniederhut
Heavily revised version of the original paper posted incrementally over the last 1+ week, now ready for a second review.

In the below I have interleaved by response to your original review.

Here is my review report: The manuscript describes a 3D audio simulator that can be used in real-time applications for cognitive hearing science. The proposed software can be useful in many aspects of hearing applications. It is within the scope of the SciPy conference and the audience can be highly interested in reading the paper and using the corresponding codes. But, the paper needs several major changes as I have noted below:

General Comment: The first concern is the language of the manuscript: There are many spelling and grammar mistakes that should be addressed. Furthermore, some statements are not clear. I have mentioned only a few of them although there are many others:

- Agree, I have made many edits

Specific Comments: Please rephrase: "To that end this simulator
is integral to planned human subject testing where headphone
testing will be used to blind test subjects for their perception of
where a sound source is located."

- Done

Use a comma and rephrase: "To produce a synthesized 3D audio sound field starts with a
geometry."

- Done; sentences reworked and the corresponding figure was reworked to make the two coordinate frames clear. Note the first frame is  

Not clear and there are two "the": " For a given source location
(x, y,z) = (x1,−x2, x3) pointing at the the origin, we transform
from rectangular coordinates to cylindrical coordinates as shown
in Figure 1."

- Done in the rework that also resolved ; What you reference above was indeed not clear, and when I went back to re-read things I concluded it was not correct 

Please use more formal English: "This transformation is motivated by [Fitzpatrick],
and will be explained more fully in a later section. "
For instance, you could say "This transformation is motivated by [Fitzpatrick],
and explained in detail in section Section Name'."

- Fixed

Do you mean 'mentioned'?: "Note also the
second rectangular coordinate frame mention above is the notation
used by CIPIC."

- Resolved; I likely had a typo here, but in my rewrite I changed the phrase to "The second system is referred to by CIPIC..."

Not clear, please check again and rephrase: "The 3D audio rendering provided by the simulator developed
in this paper, relies on the 1250 HRIR measurements were taken
using the geometrical configuration shown in Figure 2."

- Fixed; lots of rewrite in this area

Please remove 'The': "For a source on the 1 m reference sphere, or further away, the there is no ..."

- Fixed; This was an incorrect statement and is now stated correctly in this section of the paper

Please rephrase: "When the source is inside is the unit sphere sound parallax ..."

- Fixed/resolved; again this section of the paper has undergone a lot of rewrite. I have also now added Figure 6 to give a pictoral view of the parallax problem. There is not enough space to go through all of the equations behind the parallax correction, but the paper by Fitzpatrick is readily available. 

Please check for more. There are many others. I suggest a native English speaker revise the whole text.

- Resolved, I think; The first draft was written too quickly during a crunch time.

Other than the language issues, I have some other suggestions. Please see below:

The abstract is repeated in the introduction section of the main text. That is unnecessary. Please remove that part from the introduction.

- Duplication removed; this was carelessness on my part

Define CIBIC abbreviation before its first usage.

- Fixed; See the introduction. I though I had done this, but good that you caught it.

Please define the symbols of the mathematical variables such as azimuth angle, elevation,... in the text at the first usage.

- In the rewrite I have tried to make all variables clear and defined. Figure 1 for example now sets the stage for a large number of geometry related variables. I refer back to this figure later in the paper to help provide a visual as to how variables fit into the big picture. In the spherical head section some new variables get defined. The greek letter \theta gets used in multiple contexts, but the subscript is unique to the usage (I hope).

What does it mean? "For a given source location
(x, y,z) = (x1,−x2, x3) pointing at the the origin"
Please explain the equation in more detail. It is not very clear.

- Resolved; This was a confusing section. It has been rewritten and this equality no longer appears.

Use the same terminology throughout the whole manuscript: "we transform
from rectangular coordinates to cylindrical coordinates" and "convert from cylindrical coordinates
to cartesian"
Use either cartesian or rectangular, not both.

- Fixed; Cartesian is the dominant coordinate system employed, but CIPIC uses IPCS (really a spherical system, I have to interface with this) but now I try hard to make it clear that cylindrical coordinates are only used in the Jupyter notebook apps as a convenience for the user to position the audio source in what I think is a convenient way access localization when informally testing subjects.  

It is not clear what kind of analysis is performed in the section 'Spherical Head Model as a Simple Reference HRIR'. Are you reproducing any previous data here? Or comparing the results with some analytical solutions? If so, what is the error? If not, I suggest performing an error analysis of the results compared to some analytical solution.

- I am going to try to answer your questions first, then move on to justifying my approach. The references that I have provided in the spherical head model section of the paper are classics in my opinion. There are no closed-form analytical solutions for the geometries I am most interested in, but the paper by Duda and Martens has many plots and in particular my Figure 11 is very similar (not the same geometries) their Fig 6. This paper also has pseudo-code in their Appendix B, which is my function HRTF_sph() in the paper. I feel comfortable with HRIR results I have obtained. The Duda and Martens paper does provided every step they took from HRTF to HRIR, so some small differences are possible. At the end of the spherical head section I now include some comments on a listening test that was conducted.

Wrap-up: I am suggesting re-review after these changes.

- I am ready to move forward with the next round.

stargaser commented:
Hello @mwickert! I am one of the proceedings committee members, and I will complete a review of the current version, hopefully today, and definitely by tomorrow, to try to help wrap up your paper.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+.. class:: abstract
+
+   This paper describes the development of a 3D audio simulator for use in cognitive hearing science 
+   studies and also for general 3D audio experimentation. The framework that the simulator is build 
```
```suggestion
   studies and also for general 3D audio experimentation. The framework that the simulator is built 
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser replied:
(I think this change didn't make it in)

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+   devices (HAD) research and development. At its core the simulator uses digital filters to represent the 
+   sound pressure wave propagation path from the sound source to each ear canal of a human subject. 
+   Digital filters of 200 coefficients each for left and right ears are stored in a look-up table 
+   as a function of azimuth and elevation angles of the impinging sounds source.
```
```suggestion
   as a function of azimuth and elevation angles of the impinging sound's source.
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+(time-domain) or head-related transfer function (HRTF) (frequency domain) for a particular 
+human subject. 
+Traditionally human subjects are placed in an anechoic chamber with a sound 
+source placed at say one meter from the head and then the subject is moved over a range of 
```
```suggestion
source placed at e.g. one meter from the head and then the subject is moved over a range of 
```
"say" sounds too informal

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

mwickert replied:
@stargaser

I have done many small pushes, but the last one for tonight address the general comments by including a link to the HCI tutorial at UC-Davis that you have mentioned. The reference in in the introduction.

I still want to comment further on your general comments. You are correct that the HRTF/HRIR is a function of the subjects anthropometry. Our cognitive ability to localize sound is a learning process since birth. The main contribution of this paper is the apparatus for allowing human subjects to easily be evaluated for an HRFT/HRIR "fit". I am excited to move to this step in the research, but I have yet to experience this in a true clinical setting. I will have this running as a demo at the poster session.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+database of subjects, based on 27 upper torso anthropometrics (measurements) of the subject. As a simple solution, 
+we can also consider using a simple spherical head model, and its corresponding HRTF, which 
+makes use of spherical harmonics to solve for the sound pressure magnitude and phase at any location on the sphere 
+surface. A frequency sweep of magnitude and phase is then the inverse Fourier transformed to obtain the HRIR. 
```
```suggestion
surface. A frequency sweep of magnitude and phase is then inverse Fourier transformed to obtain the HRIR. 
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+The ultimate intent of the simulator is to serve as a clinical tool for blind sound source localization experiments. 
+Human subjects will be exposed to several different HRIR models, where at least one model is a *personalized 
+fit* based on deep learning using anthropometrics and/or a finite element wave equation solution using a 3D 
+rendering of the subjects shoulders and head. Note 3D rendering of a subject can be obtained using *photogrammetry*, 
```
```suggestion
rendering of the subject's shoulders and head. 3D rendering of a subject can be obtained using *photogrammetry*, 
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+===========
+
+To produce a synthesized 3D audio sound field, we start with a geometry where the center of the coordinate frame is 
+the intersection between the subjects  *mid-sagittal* or vertical *median plane* and the line 
```
```suggestion
the intersection between the subject's  *mid-sagittal* or vertical *median plane* and the line 
```

mwickert replied:
Applied in 6/30/2019 10:26 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+The primary head-centered system has cartesian coordinates labeled :math:`(x,y,z)` and associated cylindrical 
+coordinates :math:`(r_{xy},\phi_\text{az},h_y)` (black labels in Figure :ref:`CYLIND`). The cylindrical coordinates 
+will be used in Jupyter notebook apps presented later as the interface for GUI controls to conveniently position 
+the audio source about a subjects head. A secondary head-centered 
```
```suggestion
the audio source about a subject's head. A secondary head-centered 
```

mwickert replied:
Applied in 6/30/2019 10:26 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+
+In this section we briefly describe the role real-time digital signal processing (DSP) plays in implementing 
+the 3D audio simulator. A top level block diagram of the 3D audio simulator is shown in Figure :ref:`FILTERING`. 
+From the block diagram we see that for audio source is positioned at :math:`(x,y,z)` relative to the head 
```
```suggestion
For an audio source positioned at :math:`(x,y,z)` relative to the head 
```
I found the phrase "From the block diagram we see that..." confusing, because the filter coefficients, the gain scaling, the parallax, etc. are not in Figure 4 at all.

mwickert replied:
Applied in 6/30/2019 10:26 MDT push. Your comment about being confusing is good. I added this figure after the first review, because yes I think the parallax correction needs some explaining in this paper, without forcing the reader to jump to the Fitzpatrick reference. Again thanks helping me make this clearer.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+processing functions of :code:`scipy.signal` [ScipySignal]_. The 200 coefficients of the right and left HRIR 
+are equivalent to the coefficients in a finite impulse response (FIR) digital filter which produce a discrete-time 
+output signal or sequence :math:`y_R[n]/y_L[n]` from a single audio source signal :math:`x[n]`. All of the signals 
+are processed with at a sampling rate of :math:`f_s = 44.1` ksps, as this is rate used in forming the CIPIC 
```
```suggestion
are processed with a sampling rate of :math:`f_s = 44.1` kHz, as this is rate used in forming the CIPIC 
```
I couldn't parse "ksps" and I wonder if "kcps" (kilo cycles per second) was intended. Later in the paper I see 44.1 kHz so I suggest it here.

mwickert replied:
Applied in 6/30/2019 10:26 MDT push. The ksps is an abbreviation for kilo samples per second, you are correct in that just saying 44.1 kHz is consistent with the other reference to the sampling rate. Thanks for noticing this.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+corrections are addressed in detail in [Fitzpatrick]_. For a source not on the unit sphere, 
+sound parallax requires an adjustment in the HRIR coefficients, unique to the right and left ears. 
+If we extend rays from the right and left ears that pass through the sound source location and then 
+touch the unit sphere, the required azimuth values will be shifted to locations either side of the 
```
```suggestion
touch the unit sphere, the required azimuth values will be shifted to locations on either side of the 
```

mwickert replied:
Applied in 6/30/2019 10:26 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+touch the unit sphere, the required azimuth values will be shifted to locations either side of the 
+true source azimuth. The corresponding HRIR values where these rays contact the unit sphere, 
+respectively, perform the needed parallax correction. The actual database entries utilized are those 
+that are closest to the intersection point   
```
```suggestion
that are closest to the intersection points.
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+
+The 3D audio rendering provided by the simulator developed in this paper relies on the 1250 
+HRIR measurements taken using the geometrical configuration shown in Figure :ref:`CIPICLOC`. 
+A total of 45 subjects are contained in the CIPIC HRIR database, both human and the mannequin *Kemar* [CIPICHRTF]_. 
```
```suggestion
A total of 45 subjects are contained in the CIPIC HRIR database, both human and the mannequin KEMAR (Knowles Electronics Manikin for Auditory Research) [CIPICHRTF]_. 
```
(optional to spell out the acronym -- I would find it helpful)

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+Static Sound Source
+===================
+
+The first and foremost purpose the 3D audio simulator is to to be able statically position an audio source 
```
```suggestion
The first and foremost purpose the 3D audio simulator is to be able to statically position an audio source 
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+   Using spherical harmonics [Beranek]_ to calculate the pressure wave magnitude (shown here) and 
+   phase, using a plane wave audio source arriving from the bottom of the figure. :label:`SCATTER`
+
+The calculations required to obtain Figure :label:`SCATTER` follow easily using the functions found in 
```
```suggestion
The calculations required to obtain Figure :ref:`SCATTER` follow easily using the functions found in 
```

mwickert replied:
Applied in 6/30/2019 10:15 MDT push.

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,875 @@
+   for a radius 8.75 cm sphere. :label:`HRIR875`
+
+Casual listening tests with a coarse fit human subject from CIPIC and the simple spherical model, indicate both similarities 
+and differences. Coarse localization is similar between the two, bit the spherical model seems *sterile*. The fact that 
```
```suggestion
and differences. Coarse localization is similar between the two, but the spherical model seems *sterile*. The fact that 
```
I haven't encountered "sterile" used to describe a model before. Does sterile mean the HRIR is mostly flat and lacks features that a measured HRIR typically contains?

mwickert replied:
Applied in 6/30/2019 10:15 MDT push. Also added short phrase describing what is meant by sterile.

What remains is to respond to the general comments.

stargaser reviewed (changes requested):
Thank you @mwickert for this interesting submission. The paper is in good shape, with plenty of technical details and figures. I have mostly made copy-editing suggestions inline. I am marking the review "changes requested" mainly for these. I have a couple more general suggestions:

After my first time through reading the paper, I went to the CIPIC site and read through most of [the tutorial there](https://www.ece.ucdavis.edu/cipic/spatial-sound/tutorial/). My field is astronomy and I found it very helpful to skim the tutorial.

A basic question stemming from my lack of experience in your field: On my first read-through, I was puzzled about the measurements on human subjects. I was wondering if the HRIR and HRTF functions depended on an individual's perception of a sound, like: does the human subject have to signal that they heard the sound? After reading the tutorial and some descriptions on the CIPIC site, I think I understand better, that it is the location, size, and shape of the ears (pinna) that vary among human subjects.

Since your paper is getting close to the page limit and there is not much room for adding some of this background, maybe the best thing to do is to refer the reader to the tutorial.

I am hopeful that it will not take you long to address these comments.

Commit:
Applied all @Stargaser suggested changes.

Commit:
Applied all "hidden comment" suggestions.

Commit:
Added HCI tutorial link in introcution.

mwickert commented:
@stargaser

I originally made this comment under one of you smaller comments, I want it to be at the end as a general comm ent, so it is posted a second time.

I have done many small pushes, but the last one for tonight address the general comments by including a link to the HCI tutorial at UC-Davis that you have mentioned. The reference in in the introduction.

I still want to comment further on your general comments. You are correct that the HRTF/HRIR is a function of the subjects anthropometry. Our cognitive ability to localize sound is a learning process since birth. The main contribution of this paper is the apparatus for allowing human subjects to easily be evaluated for an HRFT/HRIR "fit". I am excited to move to this step in the research, but I have yet to experience this in a true clinical setting. I will have this running as a demo at the poster session.

senolpiskin commented:
Mark,
I will send my second round review today or tomorrow. I think it will be the final one since the paper looks much better now.

senolpiskin reviewed (changes requested):
The figures are of high quality and very informative. But the text needs some more editing. There are some confusing parts. I have quoted them below. Please consider rephrasing them:


"To finally render 3D audio requires selection of the appropriate
right/left filter coefficient set, and if needed range correction."
Does it mean: 'Rendering 3D audio requires a proper set of right/left filter coefficients and possibly the range correction.' ?

Not clear, please rephrase or make two sentences: " If
we simply want to position an audio source on the 1 m reference
sphere used by CIPIC, then the coefficient selection process is
simply picking the index into the database that is closest to the
corresponding IPCS angle pair, (φ,θ), of the source."

Not clear, please rephrase: "The approach
taken here follows [Fitzpatrick] using the primary cartesian coordinates to additionally perform parallax correction and source
range amplitude correction. "
Does it mean: 'Our approach follows the methodology of [Fitzpatrick] which uses the primary ...' ?

Not clear, please rephrase or make two sentences: "Source range correction accounts for
the fact that with a point source we have sound wave divergence,
thus the amplitude needs to be scaled inversely with radial distance
(inverse-squared ..."

Not clear, please rephrase or make two sentences: "The sound wave
amplitude correction factors self.tR and tL are obtained from
the parallax correction expression in [Fitzpatrick], and double as
the required range scale factors, GR and GL in (1) and (2)."

"Finally, solve
for the scattered wave, ˜ps(r,θi), by applying boundary conditions,
see [Beranek] for details, we superimpose the two solutions to
obtain
 ..."

Does it mean: 'Finally, we solve
for the scattered wave, ˜ps(r,θi), by applying boundary conditions
(see [Beranek] for details), we superimpose the two solutions to
obtain
 ...'

What does it mean the second plot coordinate? please be more specific: "The second
plot coordinate, due to axial symmetry, is w ..."

Not clear, please rephrase: "Note
in the spherical coordinates of the math model, it remains that ..."

Not clear, please rephrase or make two sentences: "The calculations required to obtain Figure 10 follow easily
using the functions found in scipy.special, e.g., for the
scattered field the calculation is:"

The sentences are not complete and vague: "Fig. 10: Using spherical harmonics [Beranek] to calculate the
pressure wave magnitude (shown here) and phase, using a plane wave
audio source arriving from the bottom of the figure."

"In signal processing the transfer function, H(f) = |H(f)|e
j∠H(f)
,
is a ratio of two complex numbers as a function frequency in Hz."
You can rewrite as: 'In signal processing, the transfer function, H(f) = |H(f)|e
j∠H(f)
,
is a ratio of two complex numbers as a function of frequency in Hz.'

Remove will: "Since
we are working with digital (discrete-time) signal processing, the
inverse discrete Fourier transform (IDFT) will is used here, as
opposed to the Fourier integral."
Does it mean: 'instead of Fourier integral'?

Please check for more, especially the sections after FIR Filter Coefficient Set Selection.

Another suggestion: The methodology can be confusing for those who are not very familiar in the field. So, would you think adding a brief flowchart of the methods section or the software that summarizes everything in a simpler way?

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,881 @@
+Static Sound Source
+===================
+
+The first and foremost purpose the 3D audio simulator is to be able to statically position an audio source 
```
```suggestion
The first and foremost purpose of the 3D audio simulator is to be able to statically position an audio source 
```

mwickert replied:
Done, will see in next push

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,881 @@
+   \frac{h_n^{(2)}(kr)}{h_n^{\prime(2)}(kR)},\ r > R
+
+where :math:`\theta_i` is the angle of incidence between the source and measurement point, 
+:math:`f` is the operating frequency in Hz, :math:`r` is the distance fro the source to the 
```
```suggestion
:math:`f` is the operating frequency in Hz, :math:`r` is the distance from the source to the 
```

mwickert replied:
Done, will see in next push

stargaser commented on papers/mark_wickert/mark_wickert.rst:
```diff
@@ -0,0 +1,881 @@
+
+where :math:`\theta_i` is the angle of incidence between the source and measurement point, 
+:math:`f` is the operating frequency in Hz, :math:`r` is the distance fro the source to the 
+center if the sphere, and one again :math:`R` is the sphere radius. Recall also that the 
```
```suggestion
center of the sphere, and once again :math:`R` is the sphere radius. Recall also that the 
```

mwickert replied:
Done, will see in next push

stargaser reviewed (commented):
@mwickert thank you for the revisions, and for including the tutorial in the introduction. I found a few more typos or missing words upon resolving the previous copy-editing suggestions I had made -- these are inline.

I agree with @senolpiskin that some of the longer sentences could be split into two or otherwise rephrased. Since the paper is very close to the page limit, it would not be possible to add an explanatory flowchart without removing other parts of the manuscript; for my part, I won't require a flowchart for labeling the paper as ready.

We are getting very close to the planned pre-conference publication of the proceedings -- if it is possible to attend to the requests for re-wording, I am hopeful we can get this wrapped up. There will also be a post-conference merge; plus the current version is available on the procbuild server for people to view.

mwickert commented:
@stargaser @senolpiskin 

I have just had a quick look at both of your comments. I have been busy trying to get two posters completed to get printed before the 4th. I agree to the sentence break-up changes, but obviously have to get into it.  hope to get working on the changes ASAP.

@stargaser if you had your wishes, when do you want me to have my pushes complete?

stargaser commented:
@mwickert understood about needing to get your posters printed. My understanding is that @deniederhut will merge the PRs labeled "Ready" on Wednesday, July 3rd, for the conference-ready version. Two weeks after that, the post-conference merge will be done.

Checking the 2018 branch, my own paper last year didn't make the pre-conference merge and was one of several merged a couple days after the conference.

If the in-line "suggestion" comments are helpful, I can make another pass of copy-editing. For those little things, it is fine just to resolve them as you fix them, and/or just put a thumbs-up emoji on the in-line comment. Or, even easier, you can use the "Commit suggestion" button. There's another button "Add suggestion to batch" to accept several of these into a single commit.

mwickert commented:
@stargaser 

Thanks for the hints. I want to do all my edits in my local version and then push. I will work on this  tonight and hopefully make the July 4th ~deadline.

stargaser commented:
My apologies -- I had originally written "Wednesday, July 4th" when looking at the 2018 calendar. I corrected the comment to say "Wednesday, July 3rd" for the pre-conference merge. (not what you wanted to hear, I'm sure...)

mwickert commented:
David,

I’m going to try after jazz gig is over this evening.

Mark Wickert

On Jul 2, 2019, at 5:22 PM, David Shupe <notifications@github.com<mailto:notifications@github.com>> wrote:


My apologies -- I had originally written "Wednesday, July 4th" when looking at the 2018 calendar. I corrected the comment to say "Wednesday, July 3rd" for the pre-conference merge. (not what you wanted to hear, I'm sure...)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/scipy-conference/scipy_proceedings/pull/481?email_source=notifications&email_token=AC74UBUH7A5FYA5K562X7FDP5PPJ3A5CNFSM4HO2O2U2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZC2B4A#issuecomment-507879664>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AC74UBULXIDIH22OLLT5UBLP5PPJ3ANCNFSM4HO2O2UQ>.

Commit:
Many small fixes applied. More to come.

mwickert commented:
@stargaser and @senolpiskin 

I have applied all the changes requested by @stargazer. I have applied fixes for some of the @senolpiskin comments. It is too late after being up let night doing posters. I will return to this in the morning.

@seolpiskin responses thus far:

"To finally render 3D audio requires selection of the appropriate
right/left filter coefficient set, and if needed range correction."
Does it mean: 'Rendering 3D audio requires a proper set of right/left filter coefficients and possibly the range correction.' ?

- Your restating my sentence is correct and my original intent. I am happy with it as is. Are you?

Not clear, please rephrase or make two sentences: " If
we simply want to position an audio source on the 1 m reference
sphere used by CIPIC, then the coefficient selection process is
simply picking the index into the database that is closest to the
corresponding IPCS angle pair, (φ,θ), of the source."

- This sentence is reworded to be simpler that my first attempt. In sourece this is lines 231-233. I have left the old text in as an RST comment

"In signal processing the transfer function, H(f) = |H(f)|e
j∠H(f)
,
is a ratio of two complex numbers as a function frequency in Hz."
You can rewrite as: 'In signal processing, the transfer function, H(f) = |H(f)|e
j∠H(f)
,
is a ratio of two complex numbers as a function of frequency in Hz.'

- Done

Remove will: "Since
we are working with digital (discrete-time) signal processing, the
inverse discrete Fourier transform (IDFT) will is used here, as
opposed to the Fourier integral."
Does it mean: 'instead of Fourier integral'?

- I have removed will

Please check for more, especially the sections after FIR Filter Coefficient Set Selection.

- Thinking about this

Another suggestion: The methodology can be confusing for those who are not very familiar in the field. So, would you think adding a brief flowchart of the methods section or the software that summarizes everything in a simpler way?

- The paper is currently very close to the 8-page limit. If I removed code listings and instead showed flowcharts, this may be helpful to some. I suspect others may find the code and the embedded comments are better than a flowchart. I am content staying with code listings. Can @stargazer please provide an opinion on this one.

Commit:
Remaining edits applied + figure mod.

mwickert commented:
@senolpiskin and @stargaser 

Remaining changes + others I have found now addressed. See the below original comments with bullets my response.

Not clear, please rephrase: "The approach
taken here follows [Fitzpatrick] using the primary cartesian coordinates to additionally perform parallax correction and source
range amplitude correction. "
Does it mean: 'Our approach follows the methodology of [Fitzpatrick] which uses the primary ...' ?

- Your Does it mean is correct. I have reworded the sentence slightly. See source lines 240-242.

Not clear, please rephrase or make two sentences: "Source range correction accounts for
the fact that with a point source we have sound wave divergence,
thus the amplitude needs to be scaled inversely with radial distance
(inverse-squared ..."

- Reworded over source lines 242-245.

Not clear, please rephrase or make two sentences: "The sound wave
amplitude correction factors self.tR and tL are obtained from
the parallax correction expression in [Fitzpatrick], and double as
the required range scale factors, GR and GL in (1) and (2)."

- Reworded over source lines 404-405. The old version is an RST in the lines 407-409

"Finally, solve
for the scattered wave, ˜ps(r,θi), by applying boundary conditions,
see [Beranek] for details, we superimpose the two solutions to
obtain
..."

Does it mean: 'Finally, we solve
for the scattered wave, ˜ps(r,θi), by applying boundary conditions
(see [Beranek] for details), we superimpose the two solutions to
obtain
...'

- Reworded over source lines 495-496.

What does it mean the second plot coordinate? please be more specific: "The second
plot coordinate, due to axial symmetry, is w ..."

- Reworded over source lines 521-524.

Not clear, please rephrase: "Note
in the spherical coordinates of the math model, it remains that ..."

- This was dealt with in the above rewording over source lines 521-524.

Not clear, please rephrase or make two sentences: "The calculations required to obtain Figure 10 follow easily
using the functions found in scipy.special, e.g., for the
scattered field the calculation is:"

- Reworded over source lines 538-539.

The sentences are not complete and vague: "Fig. 10: Using spherical harmonics [Beranek] to calculate the
pressure wave magnitude (shown here) and phase, using a plane wave
audio source arriving from the bottom of the figure."

- Figure caption reworded in source lines 545-536. Figure modified to include a direction arrow.

senolpiskin commented:
Thanks for the changes, Mark. I don't have any other comment at this point.

stargaser reviewed (approved):
This revision looks good to me. Thanks @senolpiskin for your reviews, and @mwickert for getting the changes in!

mwickert commented:
@stargaser @senolpiskin 

Thanks to the both of you for working me through this. I look forward to meeting you both next week at the conference.
